File: Calculator/web/simplify.js
/* (c) 2012, Carl Burch, cburch@cburch.com
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
simplify = (function () {
	"use strict";

	var EAGER_EVALUATION = 0;
	var LAZY_EVALUATION = 1;
	var LAZY_SLOW_EVALUATION = 2;

	var APPLIED = 0;
	var PURE_CHURCH = 1;
	var PURE = 2;

	function CandidateFinder(options) {
		this.options = options;
	}

	CandidateFinder.prototype.visitAbstract = function (e) {
		if (this.canReduce(e)) return e;
		else return e.expr.visit(this);
	};
	
	CandidateFinder.prototype.visitIdent = function (e) {
		if (this.canReduce(e)) return e;
		else return null;
	};

	CandidateFinder.prototype.canReduce = function (e) {
		// handle beta-reductions
		if (e.hasOwnProperty('left') && e.left.hasOwnProperty('expr')) {
			return true;
		}

		// handle eta-reductions
		if (this.options.useEtas && e.hasOwnProperty('expr') &&
				e.expr.hasOwnProperty('left') &&
				e.expr.right.hasOwnProperty('ident') &&
				e.expr.right.ident === e.sym &&
				!e.expr.left.uses(e.sym)) {
			return true;
		}

		// handle builtins
		if (this.options.calcType === APPLIED) {
			var funcExpr = e;
			var argCount = 0;
			while (funcExpr.hasOwnProperty('left')) {
				funcExpr = funcExpr.left;
				argCount++;
			}
			if (funcExpr.hasOwnProperty('ident') &&
					Builtin.hasOwnProperty(funcExpr.ident.name)) {
				var builtin = Builtin[funcExpr.ident.name];
				if (builtin.numArguments === argCount) {
					var args = this.getArguments(e, argCount);
					if (builtin.canApply(args)) return true;
				}
			}
		}

		return false;
	};

	CandidateFinder.prototype.reduce = function (e) {
		// handle beta-reductions
		if (e.hasOwnProperty('left') && e.left.hasOwnProperty('expr')) {
			return e.left.expr.visit(new SymbolSubstitutor(e.left.sym,
				e.right));
		}

		// handle eta-reductions
		if (this.options.useEtas && e.hasOwnProperty('expr') &&
				e.expr.hasOwnProperty('left')) {
			return e.expr.left;
		}

		// handle builtins
		if (this.options.calcType === APPLIED) {
			var funcExpr = e;
			var argCount = 0;
			while (funcExpr.hasOwnProperty('left')) {
				funcExpr = funcExpr.left;
				argCount++;
			}
			if (funcExpr.hasOwnProperty('ident') &&
					Builtin.hasOwnProperty(funcExpr.ident.name)) {
				var builtin = Builtin[funcExpr.ident.name];
				if (builtin.numArguments === argCount) {
					var args = this.getArguments(e, argCount);
					return builtin.apply(args);
				}
			}
		}

		return null;
	};

	CandidateFinder.prototype.getArguments = function (expr, argCount) {
		var argExpr = expr;
		var args = new Array(argCount);
		var argPos = argCount - 1;
		while (argExpr.hasOwnProperty('left')) {
			args[argPos] = argExpr.right;
			argPos--;
			argExpr = argExpr.left;
		}
		return args;
	};

	function LazyCandidateFinder(options) {
		this.options = options;
	}
	LazyCandidateFinder.prototype = new CandidateFinder();
	LazyCandidateFinder.prototype.constructor = LazyCandidateFinder;
	LazyCandidateFinder.prototype.visitApply = function (expr) {
		if (this.canReduce(expr)) return expr;

		var ret = expr.left.visit(this);
		if (ret !== null) return ret;
		return expr.right.visit(this);
	};

	function EagerCandidateFinder(options) {
		this.options = options;
	}
	EagerCandidateFinder.prototype = new CandidateFinder();
	EagerCandidateFinder.prototype.constructor = EagerCandidateFinder;
	EagerCandidateFinder.prototype.visitApply = function (expr) {
		var ret = expr.right.visit(this);
		if (ret !== null) return ret;

		if (this.canReduce(expr)) return expr;

		return expr.left.visit(this);
	};

	function SymbolSubstitutor(symbolToFind, exprToReplace) {
		this.toFind = symbolToFind;
		this.toReplace = exprToReplace;
	}
	SymbolSubstitutor.prototype.visitAbstract = function (e) {
		if (e.sym === this.toFind) return e;

		var expr = e.expr.visit(this);
		if (expr === e.expr) return e;
		else return new Expr.Abstract(e.sym, expr);
	};
	SymbolSubstitutor.prototype.visitApply = function (e) {
		var left = e.left.visit(this);
		var right = e.right.visit(this);
		if (left === e.left && right === e.right) return e;
		else return new Expr.Apply(left, right);
	};
	SymbolSubstitutor.prototype.visitIdent = function (e) {
		return e.ident === this.toFind ? this.toReplace : e;
	};

	function LazySubstitutor(exprSource, exprDest) {
		this.creations = new Collections.Map();
		this.source = exprSource;
		this.dest = exprDest;
	}
	LazySubstitutor.prototype.visitAbstract = function (e) {
		if (e === this.source) return this.dest;

		var expr = e.expr.visit(this);
		if (expr === e.expr) {
			return e;
		} else {
			var ret = this.creations.get(e);
			if (ret === null) {
				ret = new Expr.Abstract(e.sym, expr);
				this.creations.put(e, ret);
			}
			return ret;
		}
	};
	LazySubstitutor.prototype.visitApply = function (e) {
		if (e === this.source) return this.dest;

		var left = e.left.visit(this);
		var right = e.right.visit(this);
		if (left === e.left && right === e.right) {
			return e;
		} else {
			var ret = this.creations.get(e);
			if (ret === null) {
				ret = new Expr.Apply(left, right);
				this.creations.put(e, ret);
			}
			return ret;
		}
	};
	LazySubstitutor.prototype.visitIdent = function (e) {
		return e === this.source ? this.dest : e;
	};

	function EagerSubstitutor(exprSource, exprDest) {
		this.source = exprSource;
		this.dest = exprDest;
	}
	EagerSubstitutor.prototype.visitAbstract = function (e) {
		if (e === this.source) return this.dest;

		var expr = e.expr.visit(this);
		return expr === e.expr ? e : new Expr.Abstract(e.sym, expr);
	};
	EagerSubstitutor.prototype.visitApply = function (e) {
		if (e === this.source) return this.dest;

		var right = e.right.visit(this);
		if (right !== e.right) return new Expr.Apply(e.left, right);
		var left = e.left.visit(this);
		if (left !== e.left) return new Expr.Apply(left, e.right);
		return e;
	};
	EagerSubstitutor.prototype.visitIdent = function (e) {
		return e === this.source ? this.dest : e;
	};

	function LazySlowSubstitutor(exprSource, exprDest) {
		this.source = exprSource;
		this.dest = exprDest;
	}
	LazySlowSubstitutor.prototype.visitAbstract = function (e) {
		if (e === this.source) return this.dest;

		var expr = e.expr.visit(this);
		return expr === e.expr ? e : new Expr.Abstract(e.sym, expr);
	};
	LazySlowSubstitutor.prototype.visitApply = function (e) {
		if (e === this.source) return this.dest;

		var left = e.left.visit(this);
		if (left !== e.left) return new Expr.Apply(left, e.right);
		var right = e.right.visit(this);
		if (right !== e.right) return new Expr.Apply(e.left, right);
		return e;
	};
	LazySlowSubstitutor.prototype.visitIdent = function (e) {
		return e === this.source ? this.dest : e;
	};

	var simplify = function (original, options) {
		var candFinder;
		var substClass;
		if (options.evalOrder === EAGER_EVALUATION) {
			candFinder = new EagerCandidateFinder(options);
			substClass = EagerSubstitutor;
		} else if (options.evalOrder === LAZY_SLOW_EVALUATION) {
			candFinder = new LazyCandidateFinder(options);
			substClass = LazySlowSubstitutor;
		} else {
			candFinder = new LazyCandidateFinder(options);
			substClass = LazySubstitutor;
		}

		var source = original.visit(candFinder);
		if (source === null) return original;
		var dest = candFinder.reduce(source);
		return original.visit(new substClass(source, dest));
	};

	simplify.EAGER_EVALUATION = EAGER_EVALUATION;
	simplify.LAZY_EVALUATION = LAZY_EVALUATION;
	simplify.LAZY_SLOW_EVALUATION = LAZY_SLOW_EVALUATION;

	simplify.APPLIED = 0;
	simplify.PURE_CHURCH = PURE_CHURCH;
	simplify.PURE = PURE;

	return simplify;
}());

File: Automarker/src/automarker/Simplify.java
package automarker;

import automarker.Expr;
import java.util.HashMap;

class Simplify {

    static final int EAGER_EVALUATION = 0;
    static final int LAZY_EVALUATION = 1;
    static final int LAZY_SLOW_EVALUATION = 2;

    private static abstract class CandidateFinder implements Expr.Visitor {

        boolean use_etas;
        boolean use_applied;

        CandidateFinder(boolean paramBoolean1, boolean paramBoolean2) {
            this.use_etas = paramBoolean1;
            this.use_applied = paramBoolean2;
        }

        public Object visitAbstract(Expr.Abstract paramAbstract) {
            if (isCandidate(paramAbstract)) {
                return paramAbstract;
            }
            return paramAbstract.expr.visit(this);
        }

        public abstract Object visitApply(Expr.Apply paramApply);

        public Object visitIdent(Expr.Ident paramIdent) {
            if (isCandidate(paramIdent)) {
                return paramIdent;
            }
            return null;
        }

        boolean isCandidate(Expr paramExpr) {
            Object localObject1;
            Expr localExpr;
            if ((paramExpr instanceof Expr.Apply)) {
                localObject1 = ((Expr.Apply) paramExpr).left;
                localExpr = ((Expr.Apply) paramExpr).right;
                if ((localObject1 instanceof Expr.Abstract)) {
                    return true;
                }
            }
            Object localObject2;
            Object localObject3;
            if ((this.use_etas) && ((paramExpr instanceof Expr.Abstract))) {
                localObject1 = ((Expr.Abstract) paramExpr).sym;
                localExpr = ((Expr.Abstract) paramExpr).expr;
                if ((localExpr instanceof Expr.Apply)) {
                    localObject2 = ((Expr.Apply) localExpr).left;
                    localObject3 = ((Expr.Apply) localExpr).right;
                    if (((localObject3 instanceof Expr.Ident)) && (((Expr.Ident) localObject3).sym == localObject1) && (!((Expr) localObject2).uses((Symbol) localObject1))) {

                        return true;
                    }
                }
            }

            if (this.use_applied) {
                localObject1 = paramExpr;
                int i = 0;
                while ((localObject1 instanceof Expr.Apply)) {
                    localObject1 = ((Expr.Apply) localObject1).left;
                    i++;
                }
                if ((localObject1 instanceof Expr.Ident)) {
                    localObject2 = ((Expr.Ident) localObject1).sym.toString();
                    localObject3 = Builtin.get((String) localObject2, i);
                    if (localObject3 != null) {
                        Expr[] arrayOfExpr = new Expr[i];
                        localObject1 = paramExpr;
                        int j = i - 1;
                        while ((localObject1 instanceof Expr.Apply)) {
                            arrayOfExpr[j] = ((Expr.Apply) localObject1).right;
                            j--;
                            localObject1 = ((Expr.Apply) localObject1).left;
                        }
                        if (((Builtin) localObject3).canApply(arrayOfExpr)) {
                            return true;
                        }
                    }
                }
            }
            return false;
        }

        Expr getReplacement(Expr paramExpr) {
            Expr localExpr1;
            Expr localExpr2;
            Object localObject;
            if ((paramExpr instanceof Expr.Apply)) {
                localExpr1 = ((Expr.Apply) paramExpr).left;
                localExpr2 = ((Expr.Apply) paramExpr).right;
                if ((localExpr1 instanceof Expr.Abstract)) {
                    localObject = (Expr.Abstract) localExpr1;
                    return (Expr) ((Expr.Abstract) localObject).expr.visit(new Simplify.SymbolSubstitutor(((Expr.Abstract) localObject).sym, localExpr2));
                }
            }

            if ((this.use_etas) && ((paramExpr instanceof Expr.Abstract))) {
                localExpr1 = ((Expr.Abstract) paramExpr).expr;
                if ((localExpr1 instanceof Expr.Apply)) {
                    localExpr2 = ((Expr.Apply) localExpr1).left;
                    return localExpr2;
                }
            }

            if (this.use_applied) {
                localExpr1 = paramExpr;
                int i = 0;
                while ((localExpr1 instanceof Expr.Apply)) {
                    localExpr1 = ((Expr.Apply) localExpr1).left;
                    i++;
                }
                if ((localExpr1 instanceof Expr.Ident)) {
                    localObject = ((Expr.Ident) localExpr1).sym.toString();
                    Builtin localBuiltin = Builtin.get((String) localObject, i);
                    if (localBuiltin != null) {
                        Expr[] arrayOfExpr = new Expr[i];
                        localExpr1 = paramExpr;
                        int j = i - 1;
                        while ((localExpr1 instanceof Expr.Apply)) {
                            arrayOfExpr[j] = ((Expr.Apply) localExpr1).right;
                            j--;
                            localExpr1 = ((Expr.Apply) localExpr1).left;
                        }
                        return localBuiltin.apply(arrayOfExpr);
                    }
                }
            }

            return null;
        }
    }

    private static class LazyCandidateFinder
            extends Simplify.CandidateFinder {

        LazyCandidateFinder(boolean paramBoolean1, boolean paramBoolean2) {
            super(paramBoolean1, paramBoolean2);
        }

        public Object visitApply(Expr.Apply paramApply) {
            if (isCandidate(paramApply)) {
                return paramApply;
            }
            Object localObject = paramApply.left.visit(this);
            if (localObject != null) {
                return localObject;
            }
            return paramApply.right.visit(this);
        }
    }

    private static class EagerCandidateFinder
            extends Simplify.CandidateFinder {

        EagerCandidateFinder(boolean paramBoolean1, boolean paramBoolean2) {
            super(paramBoolean1, paramBoolean2);
        }

        public Object visitApply(Expr.Apply paramApply) {
            Object localObject = paramApply.right.visit(this);
            if (localObject != null) {
                return localObject;
            }
            if (isCandidate(paramApply)) {
                return paramApply;
            }
            return paramApply.left.visit(this);
        }
    }

    private static class SymbolSubstitutor implements Expr.Visitor {

        Symbol to_find;
        Expr to_replace;

        SymbolSubstitutor(Symbol paramSymbol, Expr paramExpr) {
            this.to_find = paramSymbol;
            this.to_replace = paramExpr;
        }

        public Object visitAbstract(Expr.Abstract paramAbstract) {
            if (paramAbstract.sym == this.to_find) {
                return paramAbstract;
            }
            Expr localExpr = (Expr) paramAbstract.expr.visit(this);
            return localExpr == paramAbstract.expr ? paramAbstract : new Expr.Abstract(paramAbstract.sym, localExpr);
        }

        public Object visitApply(Expr.Apply paramApply) {
            Expr localExpr1 = (Expr) paramApply.left.visit(this);
            Expr localExpr2 = (Expr) paramApply.right.visit(this);
            if ((localExpr1 == paramApply.left) && (localExpr2 == paramApply.right)) {
                return paramApply;
            }
            return new Expr.Apply(localExpr1, localExpr2);
        }

        public Object visitIdent(Expr.Ident paramIdent) {
            return paramIdent.sym == this.to_find ? this.to_replace : paramIdent;
        }
    }

    private static class LazySubstitutor implements Expr.Visitor {

        HashMap creations = new HashMap();
        Expr source;
        Expr dest;

        LazySubstitutor(Expr paramExpr1, Expr paramExpr2) {
            this.source = paramExpr1;
            this.dest = paramExpr2;
        }

        public Object visitAbstract(Expr.Abstract paramAbstract) {
            if (paramAbstract == this.source) {
                return this.dest;
            }
            Expr localExpr = (Expr) paramAbstract.expr.visit(this);
            if (localExpr == paramAbstract.expr) {
                return paramAbstract;
            }
            Object localObject = (Expr) this.creations.get(paramAbstract);
            if (localObject == null) {
                localObject = new Expr.Abstract(paramAbstract.sym, localExpr);
                this.creations.put(paramAbstract, localObject);
            }
            return localObject;
        }

        public Object visitApply(Expr.Apply paramApply) {
            if (paramApply == this.source) {
                return this.dest;
            }
            Expr localExpr1 = (Expr) paramApply.left.visit(this);
            Expr localExpr2 = (Expr) paramApply.right.visit(this);
            if ((localExpr1 == paramApply.left) && (localExpr2 == paramApply.right)) {
                return paramApply;
            }
            Object localObject = (Expr) this.creations.get(paramApply);
            if (localObject == null) {
                localObject = new Expr.Apply(localExpr1, localExpr2);
                this.creations.put(paramApply, localObject);
            }
            return localObject;
        }

        public Object visitIdent(Expr.Ident paramIdent) {
            return paramIdent == this.source ? this.dest : paramIdent;
        }
    }

    private static class EagerSubstitutor implements Expr.Visitor {

        Expr source;
        Expr dest;

        EagerSubstitutor(Expr paramExpr1, Expr paramExpr2) {
            this.source = paramExpr1;
            this.dest = paramExpr2;
        }

        public Object visitAbstract(Expr.Abstract paramAbstract) {
            if (paramAbstract == this.source) {
                return this.dest;
            }
            Expr localExpr = (Expr) paramAbstract.expr.visit(this);
            return localExpr == paramAbstract.expr ? paramAbstract : new Expr.Abstract(paramAbstract.sym, localExpr);
        }

        public Object visitApply(Expr.Apply paramApply) {
            if (paramApply == this.source) {
                return this.dest;
            }
            Expr localExpr1 = (Expr) paramApply.right.visit(this);
            if (localExpr1 != paramApply.right) {
                return new Expr.Apply(paramApply.left, localExpr1);
            }
            Expr localExpr2 = (Expr) paramApply.left.visit(this);
            if (localExpr2 != paramApply.left) {
                return new Expr.Apply(localExpr2, paramApply.right);
            }
            return paramApply;
        }

        public Object visitIdent(Expr.Ident paramIdent) {
            return paramIdent == this.source ? this.dest : paramIdent;
        }
    }

    private static class LazySlowSubstitutor implements Expr.Visitor {

        Expr source;
        Expr dest;

        LazySlowSubstitutor(Expr paramExpr1, Expr paramExpr2) {
            this.source = paramExpr1;
            this.dest = paramExpr2;
        }

        public Object visitAbstract(Expr.Abstract paramAbstract) {
            if (paramAbstract == this.source) {
                return this.dest;
            }
            Expr localExpr = (Expr) paramAbstract.expr.visit(this);
            return localExpr == paramAbstract.expr ? paramAbstract : new Expr.Abstract(paramAbstract.sym, localExpr);
        }

        public Object visitApply(Expr.Apply paramApply) {
            if (paramApply == this.source) {
                return this.dest;
            }
            Expr localExpr1 = (Expr) paramApply.left.visit(this);
            if (localExpr1 != paramApply.left) {
                return new Expr.Apply(localExpr1, paramApply.right);
            }
            Expr localExpr2 = (Expr) paramApply.right.visit(this);
            if (localExpr2 != paramApply.right) {
                return new Expr.Apply(paramApply.left, localExpr2);
            }
            return paramApply;
        }

        public Object visitIdent(Expr.Ident paramIdent) {
            return paramIdent == this.source ? this.dest : paramIdent;
        }
    }

    static Expr simplify(Expr paramExpr) {
        boolean bool1 = Options.getEtaReductionsOption().getValue();
        boolean bool2 = Options.getUseAppliedOption().getValue();
        int i = Options.getEvaluationOrderOption().getValue();
        Object localObject1;
        if (i == 0) {
            localObject1 = new EagerCandidateFinder(bool1, bool2);
        } else {
            localObject1 = new LazyCandidateFinder(bool1, bool2);
        }

        Expr localExpr1 = (Expr) paramExpr.visit((Expr.Visitor) localObject1);
        if (localExpr1 == null) {
            return paramExpr;
        }
        Expr localExpr2 = ((CandidateFinder) localObject1).getReplacement(localExpr1);

        Object localObject2;
        if (i == 0) {
            localObject2 = new EagerSubstitutor(localExpr1, localExpr2);
        } else if (i == 2) {
            localObject2 = new LazySlowSubstitutor(localExpr1, localExpr2);
        } else {
            localObject2 = new LazySubstitutor(localExpr1, localExpr2);
        }
        return (Expr) paramExpr.visit((Expr.Visitor) localObject2);
    }
}
